text/0000-else-match.md

```diff
@@ -0,0 +1,203 @@
+Before:
+
+```rust
+if class == self.class {
+  match rr.get_rr_type() {
+    RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+    _ => (),
+  }
+} else {
+  match class {
+    DNSClass::ANY => {
+      if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+      if let &RData::NULL(..) = rr.get_rdata() { () }
+      else { return Err(ResponseCode::FormErr) }
+      match rr.get_rr_type() {
+        RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+        _ => (),
+      }
+    },
+    DNSClass::NONE => {
+      if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+      match rr.get_rr_type() {
+        RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+        _ => (),
+      }
+    },
+    _ => return Err(ResponseCode::FormErr),
+  }
+}
+```
+
+After:
+
+```rust
+if class == self.class {
+  match rr.get_rr_type() {
+    RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+    _ => (),
+  }
+} else match class {
+  DNSClass::ANY => {
+    if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+    if let &RData::NULL(..) = rr.get_rdata() { () }
+    else { return Err(ResponseCode::FormErr) }
+    match rr.get_rr_type() {
+      RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+      _ => (),
+    }
+  },
+  DNSClass::NONE => {
+    if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+    match rr.get_rr_type() {
+      RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+      _ => (),
+    }
+  },
+  _ => return Err(ResponseCode::FormErr),
+}
+```
+
+# Detailed design
+[design]: #detailed-design
+
+## Grammar
+
+See the following document for an (incomplete) guide to the grammar used in Rust:
+[Rust Documentation → Grammar](https://doc.rust-lang.org/grammar.html).
+
+This proposal modifies the
+[if expression grammar](https://doc.rust-lang.org/grammar.html#if-expressions).
+
+```
+ else_tail : "else" [ if_expr | if_let_expr
++                   | match_expr
+                    | '{' block '}' ] ;
+```
+
+## Execution
+
+An `else match` block should be treated exactly the same as an `else` block with a single `match`
+expression.
+
+### Dead code
+
+The next `else` block after an `else match` is never run. This is because `match` itself does not
+take on an `else` block. Whether `match` should allow an `else` or not should be addressed in a
+separate proposal.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Slight maintainability problems whenever you need to add additional logic to an `else` block.
+- Can be more of a stylistic issue than an ergonomics issue.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Not an alternative but an addition to the proposal: `if match` expressions. This would add an
+additional grammar rule and modify an existing one:
+
+```
++ if_match_expr : "if" match_expr else_tail ? ;
```

Comment:
I'm trying to understand the semantics of the construct being proposed in this alternative.

Is the idea of this alternative extended proposal that the new `if match` form would provide a way to describe a (non-exhaustive) match that has an (non-optional) `else` attached? (And if the given match is exhaustive, then the attached else becomes optional and in fact pointless?)

E.g. are you suggesting something like this:

```rust
enum Cases { Case1, Case2, Case3 }
if match Case2 { Case1 => 1, Case2 => 2 } else { 3 }; // => 2
if match Case3 { Case1 => 1, Case2 => 2 } else { 3 }; // => 3
```

I know this is only an alternative so it may not seem as important to tie down the described semantics; but I would like to understand the proposal so that I can evaluate the main text in the appropriate context.